docs/elasticsearch_statefulset.md

```diff
@@ -0,0 +1,215 @@
+            'awesomeness': 42
+        })
+
+
+    if __name__ == "__main__":
+        populateData(es_conn)
+
+In [run.py](../app/run.py) we add another resource which will get the message from elasticsearch. 
+
+    class ElasticsearchResource(object):
+        def on_get(self, req, resp):
+            resp.status = falcon.HTTP_200
+            data = es_conn.get(index='messages', doc_type='song', id=1)
+            resp.body = json.dumps(data['_source'])
+
+    if es_conn:
+      app.add_route('/app/elasticsearch', ElasticsearchResource())
+ 
+### Application initContainers
+Finally, we want to make sure that our app doesn't start before the elasticsearch service is ready. We add two initContainers to check and populate elasticsearch in [\_elasticsearch_init.tpl.yaml](../helm/aladdin-demo/templates/_elasticsearch_init.tpl.yaml).
+
+    {{ define "elasticsearch_check" -}}
+    {{ if .Values.elasticsearch.create }}
+    - name: {{ .Chart.Name }}-elasticsearch-check
+      image: byrnedo/alpine-curl
+      command:
+      - 'sh'
+      - '-c'
+      - 'until curl {{ .Chart.Name }}-elasticsearch:{{ .Values.elasticsearch.port }}; do echo waiting for elasticsearch pod; sleep 2; done;'
+    {{ end }}
+    {{ end }}
+
+    ---
+
+    {{ define "elasticsearch_populate" -}}
+    {{ if .Values.elasticsearch.create }}
+    - name: {{ .Chart.Name }}-elasticsearch-populate
+      image: {{ .Values.deploy.ecr }}{{ .Chart.Name }}:{{ .Values.deploy.imageTag }}
+      command:
+      - 'python3'
+      - 'elasticsearch_util/elasticsearch_populate.py'
+      envFrom:
+        - configMapRef:
+            name: {{ .Chart.Name }}
+    {{ end }}
+    {{ end }}
+
+We include these initContainers in [aladdin-demo.deploy.yaml](../helm/aladdin-demo/templates/aladdin-demo.deploy.yaml). Since we are using persistent volume for data storage, we should only need to populate the data once, so the population initContainer is controlled by a `elasticsearch.populate` boolean specified in [values.yaml](../heml/aladdin-demo/values.yaml).
+
+## Test Persistency
+Make sure that `elasticsearch.create` and `elasticsearch.populate` are set to `true` in [values.yaml](../heml/aladdin-demo/values.yaml). Build and deploy aladdin-demo and wait until all the pods are running.
+
+    $ aladdin build
+    $ aladdin start
+    $ kubectl get pods
+
+Check that elasticsearch has been populated by curling the appropriate endpoint of the aladdin-demo service.
+
+    $ curl $(minikube service --url aladdin-demo)/app/elasticsearch
+    
+    Data from ElasticSearch is {"author": "Aladdin", "song": "A Whole New World", "lyrics": ["I can show you the world"], "awesomeness": 42} 
+    
+Now set `elasticsearch.populate` to `false` and redeploy the app with 
+   
+    $ aladdin restart
+Check to see if the elasticsearch still works, even though we didn't populate it this time around! You should get the same output as last time, meaning the data persisted between deployments, hooray!
```

Comment:
Since it is not documented anywhere and it is a big issue, we should also talk about affinity.
i.e. lets say EBS volume is in one zone 1a. k8s has no idea about it, so it will randomly create the pods and try to connect. This will fail. Hence affinity should be specified in deployment yaml. This prevents k8s from doing zone failover.
Alternative is to create EFS volume, which enables zone failover but has slower speeds and higher costs.
Third alternative is to use GlusterFS or some other distributed file storage mechanism. This enables zone failover and speed issues but it adds complexity in creating and maintaining an env.

Reply:
I added an explanation in the readme, just under the code snippet on `volumeClaimTemplates`